extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -538,6 +567,179 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
     }
   }
 
+  /**
+   * Sends RTC request as specified by rtcConfig. Returns promise which
+   * resolves to the time that the RTC callout took to complete.
+   * If disableStalewhilerevalidate == true, then we will only
+   * send one RTC request per page. If it's !== true, then we always
+   * send two requests. The first will try
+   * to hit browser cache, and the second request will always bypass.
+   * This is a way for us to implement a simple stale-while-revalidate
+   * model.
+   * The targeting info from the RTC updates the targeting info on
+   * this object within mergeRtc.
+   * @return {?Promise<?Object>} An object of parameters to add to
+   *   the ad request url.
+   * @private
+   */
+  executeRtc_() {
+    if (rtcPromise) {
+      return this.mergeRtc();
+    }
+    const ampRtcPageElement = document.getElementById('amp-rtc');
+    if (!ampRtcPageElement) {
+      return Promise.resolve();
+    }
+    let endpoint;
+    rtcConfig = tryParseJson(ampRtcPageElement.textContent);
+    if (!isObject(rtcConfig) || !(endpoint = rtcConfig['endpoint']) ||
+        typeof endpoint != 'string' || !isSecureUrl(endpoint)) {
+      user().warn(TAG, 'Sending ad request without RTC callout,' +
+          `invalid RTC config: ${ampRtcPageElement.textContent}`);
+      return Promise.resolve();
+    }
+    let rtcTotalTime;
+    /*
+     *  disableSWR should be set to true if the endpoint is not
+     *  returning cache headers.
+     */
+    const disableSWR = (
+        rtcConfig['disableStaleWhileRevalidate'] === true);
+    if (rtcConfig['disableStaleWhileRevalidate'] != undefined &&
+    typeof rtcConfig['disableStaleWhileRevalidate'] != 'boolean') {
+      user().warn('RTC disableStaleWhileRevalidate must be a boolean.');
+    }
+    const headers = new Headers();
+    headers.append('Cache-Control', 'max-age=0');
+    const init = {
+      credentials: 'include',
+    };
+    // If stale-while-revalidate is disabled, our first RTC callout
```

Comment:
Wait why?  I think my last comment might have been confusing.  The flow should be:
- always send RTC with ability to read from cache
- if response indicates not to use SWR OR response indicates it was not cached, then do NOT make SWR.  The latter might be difficult depending on if you can read cache headers clientside